// @cdklabs/generative-ai-cdk-constructs/lib/patterns/gen-ai/aws-model-deployment-sagemaker/container-image.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContainerImage = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const assets = require("aws-cdk-lib/aws-ecr-assets");
const helpers_internal_1 = require("aws-cdk-lib/core/lib/helpers-internal");
/**
 *
 * https://github.com/aws/deep-learning-containers/blob/master/available_images.md
 */
class ContainerImage {
    static fromEcrRepository(repository, tag = 'latest') {
        return new EcrImage(repository, tag);
    }
    static fromAsset(directory, options = {}) {
        return new AssetImage(directory, options);
    }
}
exports.ContainerImage = ContainerImage;
_a = JSII_RTTI_SYMBOL_1;
ContainerImage[_a] = { fqn: "@cdklabs/generative-ai-cdk-constructs.ContainerImage", version: "0.1.309" };
class EcrImage extends ContainerImage {
    constructor(repository, tag) {
        super();
        this.repository = repository;
        this.tag = tag;
    }
    bind(_scope, grantable) {
        this.repository.grantPull(grantable);
        return {
            imageName: this.repository.repositoryUriForTag(this.tag),
        };
    }
}
class AssetImage extends ContainerImage {
    constructor(directory, options = {}) {
        super();
        this.directory = directory;
        this.options = options;
    }
    bind(scope, grantable) {
        // Retain the first instantiation of this asset
        if (!this.asset) {
            this.asset = new assets.DockerImageAsset(scope, `ModelImage${(0, helpers_internal_1.md5hash)(this.directory)}`, {
                directory: this.directory,
                ...this.options,
            });
        }
        this.asset.repository.grantPull(grantable);
        return {
            imageName: this.asset.imageUri,
        };
    }
}
