// @cdklabs/generative-ai-cdk-constructs/lib/cdk-lib/bedrock/knowledge-bases/supplemental-data-storage.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SupplementalDataStorageLocation = exports.SupplementalDataStorageLocationType = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
var SupplementalDataStorageLocationType;
(function (SupplementalDataStorageLocationType) {
    /**
       * Contains information about the Amazon S3 location for the extracted images.
       */
    SupplementalDataStorageLocationType["S3"] = "S3";
})(SupplementalDataStorageLocationType || (exports.SupplementalDataStorageLocationType = SupplementalDataStorageLocationType = {}));
/**
 * Represents a supplemental data storage location for images extracted from multimodal documents in your data source.
 */
class SupplementalDataStorageLocation {
    /**
       * Creates a new S3 supplemental data storage location
       * @param config The configuration for the storage location
       * @returns A new SupplementalDataStorageLocation instance
       */
    static s3(config) {
        return new SupplementalDataStorageLocation(SupplementalDataStorageLocationType.S3, config);
    }
    /**
       * Creates a new SupplementalDataStorageLocation
       * @param type The type of the storage location
       * @param locationConfig The configuration for the storage location
       */
    constructor(type, locationConfig) {
        this.type = type;
        this.locationConfig = locationConfig;
    }
    /**
       * Renders as Cfn Property
       * @internal This is an internal core function and should not be called directly.
       */
    __render() {
        if (this.type === SupplementalDataStorageLocationType.S3) {
            const s3Config = this.locationConfig;
            return {
                supplementalDataStorageLocationType: this.type,
                s3Location: {
                    uri: s3Config.uri,
                },
            };
        }
        // This should never happen as we only support S3 for now,
        // but TypeScript requires a return statement for all code paths
        throw new Error(`Unsupported storage location type: ${this.type}`);
    }
}
exports.SupplementalDataStorageLocation = SupplementalDataStorageLocation;
_a = JSII_RTTI_SYMBOL_1;
SupplementalDataStorageLocation[_a] = { fqn: "@cdklabs/generative-ai-cdk-constructs.bedrock.SupplementalDataStorageLocation", version: "0.1.309" };
