// @cdklabs/generative-ai-cdk-constructs/lib/cdk-lib/bedrock/data-sources/context-enrichment.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ContextEnrichment = exports.EnrichmentStrategyConfigurationType = exports.ContextEnrichmentType = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_iam_1 = require("aws-cdk-lib/aws-iam");
/**
 * Enum representing the type of context enrichment.
 */
var ContextEnrichmentType;
(function (ContextEnrichmentType) {
    /**
     * Uses a Bedrock Foundation Model for context enrichment.
     */
    ContextEnrichmentType["BEDROCK_FOUNDATION_MODEL"] = "BEDROCK_FOUNDATION_MODEL";
})(ContextEnrichmentType || (exports.ContextEnrichmentType = ContextEnrichmentType = {}));
/**
 * Enum representing the method to be used for enrichment strategy.
 */
var EnrichmentStrategyConfigurationType;
(function (EnrichmentStrategyConfigurationType) {
    EnrichmentStrategyConfigurationType["CHUNK_ENTITY_EXTRACTION"] = "CHUNK_ENTITY_EXTRACTION";
})(EnrichmentStrategyConfigurationType || (exports.EnrichmentStrategyConfigurationType = EnrichmentStrategyConfigurationType = {}));
/**
 * Abstract class representing a context enrichment strategy.
 * The enrichment stategy used to provide additional context.
 * For example, Neptune GraphRAG uses Amazon Bedrock foundation
 * models to perform chunk entity extraction.
 */
class ContextEnrichment {
    // ------------------------------------------------------
    // FM Enrichment Strategy
    // ------------------------------------------------------
    /**
     * Creates a Foundation Model-based enrichment strategy used to provide additional context
     * to the RAG application.
     */
    static foundationModel(props) {
        class FoundationModelContextEnrichment extends ContextEnrichment {
            constructor() {
                super(...arguments);
                this.configuration = {
                    type: ContextEnrichmentType.BEDROCK_FOUNDATION_MODEL,
                    bedrockFoundationModelConfiguration: {
                        modelArn: props.enrichmentModel.invokableArn,
                        enrichmentStrategyConfiguration: {
                            method: EnrichmentStrategyConfigurationType.CHUNK_ENTITY_EXTRACTION,
                        },
                    },
                };
            }
            generatePolicyStatements() {
                return [
                    new aws_iam_1.PolicyStatement({
                        actions: ['bedrock:InvokeModel*'],
                        resources: [props.enrichmentModel.invokableArn],
                    }),
                ];
            }
        }
        return new FoundationModelContextEnrichment();
    }
}
exports.ContextEnrichment = ContextEnrichment;
_a = JSII_RTTI_SYMBOL_1;
ContextEnrichment[_a] = { fqn: "@cdklabs/generative-ai-cdk-constructs.bedrock.ContextEnrichment", version: "0.1.309" };
