// @cdklabs/generative-ai-cdk-constructs/lib/cdk-lib/mongodb-atlas/mongodb-atlas-vector-store.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MongoDBAtlasVectorStore = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 * Construct for MongoDB Atlas vector store
 */
class MongoDBAtlasVectorStore {
    /**
     * Creates a new instance of the MongoDBAtlas class
     *
     * @param props - The properties for the MongoDB Atlas vector store
     */
    constructor(props) {
        this.collectionName = props.collectionName;
        this.credentialsSecretArn = props.credentialsSecretArn;
        this.databaseName = props.databaseName;
        this.endpoint = props.endpoint;
        this.endpointServiceName = props.endpointServiceName;
        this.fieldMapping = props.fieldMapping;
        this.vectorIndexName = props.vectorIndexName;
        this.textIndexName = props.textIndexName;
    }
}
exports.MongoDBAtlasVectorStore = MongoDBAtlasVectorStore;
_a = JSII_RTTI_SYMBOL_1;
MongoDBAtlasVectorStore[_a] = { fqn: "@cdklabs/generative-ai-cdk-constructs.mongodbAtlas.MongoDBAtlasVectorStore", version: "0.1.309" };
