// @cdklabs/generative-ai-cdk-constructs/lib/cdk-lib/neptune/graph.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NeptuneGraph = exports.NeptuneGraphBase = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
const aws_cdk_lib_1 = require("aws-cdk-lib");
const aws_cloudwatch_1 = require("aws-cdk-lib/aws-cloudwatch");
const iam = require("aws-cdk-lib/aws-iam");
const neptunegraph = require("aws-cdk-lib/aws-neptunegraph");
const notebook_1 = require("./notebook");
const utils_1 = require("../../common/helpers/utils");
/******************************************************************************
 *                            BASE CLASS
 *****************************************************************************/
/**
 * Abstract base class for a Neptune Graph.
 * Contains methods and attributes valid for Graphs either created with CDK or imported.
 */
class NeptuneGraphBase extends aws_cdk_lib_1.Resource {
    /**
     * Grant the given principal identity permissions to perform actions on this agent alias.
     */
    grant(grantee, actions) {
        return iam.Grant.addToPrincipal({
            grantee,
            actions,
            resourceArns: [this.graphArn],
            scope: this,
        });
    }
    /**
     * Grant the given identity full access to the Graph.
     */
    grantFullAccess(grantee) {
        return this.grant(grantee, ['neptune-graph:*']);
    }
    /**
     * Grant the given identity the permissions to query the Graph.
     */
    grantQuery(grantee) {
        return this.grant(grantee, [
            'neptune-graph:GetGraph',
            'neptune-graph:ReadDataViaQuery',
            'neptune-graph:WriteDataViaQuery',
            'neptune-graph:DeleteDataViaQuery',
            'neptune-graph:GetQueryStatus',
            'neptune-graph:CancelQuery',
        ]);
    }
    /**
     * Grant the given identity the permissions to read the Graph.
     */
    grantReadOnly(grantee) {
        return this.grant(grantee, [
            'neptune-graph:Get*',
            'neptune-graph:List*',
            'neptune-graph:Read*',
        ]);
    }
    /**
     * Grant the given identity the permissions to export the graph into columnar structured .csv and .parquet files.
     */
    grantExportTask(grantee) {
        return this.grant(grantee, [
            'neptune-graph:GetExportTask',
            'neptune-graph:ListExportTasks',
            'neptune-graph:StartExportTask',
            'neptune-graph:CancelExportTask',
        ]);
    }
    /**
     * Return the given named metric for this graph.
     *
     * By default, the metric will be calculated as a sum over a period of 5 minutes.
     * You can customize this by using the `statistic` and `period` properties.
     */
    metric(metricName, props) {
        const metricProps = {
            namespace: 'AWS/Neptune',
            metricName,
            dimensionsMap: { GraphIdentifier: this.graphId },
            ...props,
        };
        return this.configureMetric(metricProps);
    }
    /**
     * Returns metric for number of queued requests per second
     */
    metricNumQueuedRequestsPerSec(props) {
        return this.metric('NumQueuedRequestsPerSec', props);
    }
    /**
     * Returns metric for number of OpenCypher requests per second
     */
    metricNumOpenCypherRequestsPerSec(props) {
        return this.metric('NumOpenCypherRequestsPerSec', props);
    }
    /**
     * Returns metric for number of OpenCypher client errors per second
     */
    metricNumOpenCypherClientErrorsPerSec(props) {
        return this.metric('NumOpenCypherClientErrorsPerSec', props);
    }
    /**
     * Returns metric for number of OpenCypher server errors per second
     */
    metricNumOpenCypherServerErrorsPerSec(props) {
        return this.metric('NumOpenCypherServerErrorsPerSec', props);
    }
    /**
     * Returns metric for number of edge properties
     */
    metricNumEdgeProperties(props) {
        return this.metric('NumEdgeProperties', props);
    }
    /**
     * Returns metric for number of edges
     */
    metricNumEdges(props) {
        return this.metric('NumEdges', props);
    }
    /**
     * Returns metric for CPU utilization
     */
    metricCPUUtilization(props) {
        return this.metric('CPUUtilization', props);
    }
    /**
     * Returns metric for number of vectors
     */
    metricNumVectors(props) {
        return this.metric('NumVectors', props);
    }
    /**
     * Returns metric for number of vertex properties
     */
    metricNumVertexProperties(props) {
        return this.metric('NumVertexProperties', props);
    }
    /**
     * Returns metric for graph size in bytes
     */
    metricGraphSizeBytes(props) {
        return this.metric('GraphSizeBytes', props);
    }
    /**
     * Returns metric for graph storage usage percentage
     */
    metricGraphStorageUsagePercent(props) {
        return this.metric('GraphStorageUsagePercent', props);
    }
    configureMetric(props) {
        return new aws_cloudwatch_1.Metric({
            ...props,
            region: props?.region ?? this.stack.region,
            account: props?.account ?? this.stack.account,
        });
    }
    /**
     * Creates a Neptune Graph Notebook for the graph. Defaults to a ml.t3.medium instance type.
     * **Note: Creating a notebook will incur additional AWS costs for the notebook instance.**
     */
    createNotebook(params) {
        return new notebook_1.NeptuneGraphNotebook(this, 'Notebook', {
            ...params,
            graph: this,
        });
    }
}
exports.NeptuneGraphBase = NeptuneGraphBase;
_a = JSII_RTTI_SYMBOL_1;
NeptuneGraphBase[_a] = { fqn: "@cdklabs/generative-ai-cdk-constructs.neptune.NeptuneGraphBase", version: "0.1.309" };
/******************************************************************************
 *                        NEW CONSTRUCT DEFINITION
 *****************************************************************************/
/**
 * Class to create an L2 Neptune Analytics Graph with CDK.
 * @cloudformationResource AWS::NeptuneGraph::Graph
 */
class NeptuneGraph extends NeptuneGraphBase {
    // ------------------------------------------------------
    // Import Methods
    // ------------------------------------------------------
    static fromGraphId(scope, id, graphId) {
        class Import extends NeptuneGraphBase {
            constructor() {
                super(...arguments);
                this.graphId = graphId;
                this.graphArn = aws_cdk_lib_1.Stack.of(scope).formatArn({
                    service: 'neptune-graph',
                    resource: 'graph',
                    resourceName: graphId,
                    arnFormat: aws_cdk_lib_1.ArnFormat.SLASH_RESOURCE_NAME,
                });
                this.graphEndpoint = `${this.graphId}.${aws_cdk_lib_1.Stack.of(scope).region}.neptune-graph.amazonaws.com`;
            }
        }
        return new Import(scope, id);
    } // ------------------------------------------------------
    constructor(scope, id, props) {
        super(scope, id);
        // ------------------------------------------------------
        // Set properties or defaults
        // ------------------------------------------------------
        this.provisionedMemory = props.provisionedMemoryNCUs ?? 16;
        this.graphName =
            props.graphName ??
                (0, utils_1.generatePhysicalNameV2)(this, 'bedrock-kb-graph', {
                    separator: '-',
                    maxLength: 63,
                    lower: true,
                });
        this.publicConnectivity = props.publicConnectivity ?? true;
        this.replicaCount = props.replicaCount ?? 0;
        // ------------------------------------------------------
        // L1 Instantiation
        // ------------------------------------------------------
        this._resource = new neptunegraph.CfnGraph(this, 'Resource', {
            provisionedMemory: this.provisionedMemory,
            graphName: this.graphName,
            publicConnectivity: this.publicConnectivity,
            replicaCount: this.replicaCount,
            vectorSearchConfiguration: props.vectorSearchDimension
                ? {
                    vectorSearchDimension: props.vectorSearchDimension,
                }
                : undefined,
            deletionProtection: props.deletionProtection ?? false,
        });
        this.graphArn = this._resource.attrGraphArn;
        this.graphId = this._resource.attrGraphId;
        this.graphEndpoint = this._resource.attrEndpoint;
    }
}
exports.NeptuneGraph = NeptuneGraph;
_b = JSII_RTTI_SYMBOL_1;
NeptuneGraph[_b] = { fqn: "@cdklabs/generative-ai-cdk-constructs.neptune.NeptuneGraph", version: "0.1.309" };
