// @cdklabs/generative-ai-cdk-constructs/lib/patterns/gen-ai/aws-bedrock-batch-stepfn/create-model-invocation-job-function.js
"use strict";
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateModelInvocationJobFunction = void 0;
const aws_cdk_lib_1 = require("aws-cdk-lib");
const cdk_nag_1 = require("cdk-nag");
const bedrock_batch_sfn_lambda_code_1 = require("./bedrock-batch-sfn-lambda-code");
/**
 * Lambda function that creates Bedrock model invocation jobs for a Step Functions wait for callback task
 */
class CreateModelInvocationJobFunction extends aws_cdk_lib_1.aws_lambda.Function {
    constructor(scope, id, props) {
        super(scope, id, {
            code: bedrock_batch_sfn_lambda_code_1.BedrockBatchSfnLambdaCode.getOrCreate(),
            handler: 'aws_bedrock_batch_sfn.bedrock_create_model_invocation_job.handler',
            runtime: aws_cdk_lib_1.aws_lambda.Runtime.PYTHON_3_13,
            architecture: aws_cdk_lib_1.aws_lambda.Architecture.ARM_64,
            memorySize: 128,
            timeout: aws_cdk_lib_1.Duration.seconds(30),
        });
        this.addToRolePolicy(new aws_cdk_lib_1.aws_iam.PolicyStatement({
            actions: ['bedrock:CreateModelInvocationJob', 'bedrock:TagResource'],
            resources: [
                aws_cdk_lib_1.Stack.of(this).formatArn({
                    service: 'bedrock',
                    resource: 'model-invocation-job',
                    resourceName: '*',
                }),
            ],
        }));
        this.role?.addManagedPolicy(props.bedrockBatchPolicy);
        cdk_nag_1.NagSuppressions.addResourceSuppressions(this, [
            {
                id: 'AwsSolutions-IAM4',
                reason: 'Allow the use of AWS Managed Policies for Lambda execution.',
                appliesTo: [
                    'Policy::arn:<AWS::Partition>:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole',
                ],
            },
            {
                id: 'AwsSolutions-IAM5',
                reason: 'Wildcards allow create of Bedrock model invocation jobs.',
            },
        ], true);
    }
}
exports.CreateModelInvocationJobFunction = CreateModelInvocationJobFunction;
