// @cdklabs/generative-ai-cdk-constructs/lib/cdk-lib/bedrock/inference-profiles/common.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.InferenceProfileBase = exports.InferenceProfileType = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
const aws_cdk_lib_1 = require("aws-cdk-lib");
const aws_iam_1 = require("aws-cdk-lib/aws-iam");
/** These are the values used by the API when using aws bedrock get-inference-profile --inference-profile-identifier XXXXXXX */
var InferenceProfileType;
(function (InferenceProfileType) {
    /**
     * An inference profile that is created by AWS. These are profiles such as cross-region
     * which help you distributed traffic across a geographic region.
     */
    InferenceProfileType["SYSTEM_DEFINED"] = "SYSTEM_DEFINED";
    /**
     * An inference profile that is user-created. These are profiles that help
     * you track costs or metrics.
     */
    InferenceProfileType["APPLICATION"] = "APPLICATION";
})(InferenceProfileType || (exports.InferenceProfileType = InferenceProfileType = {}));
/**
 * Abstract base class for a ApplicationInferenceProfile.
 * Contains methods and attributes valid for ApplicationInferenceProfiles either created with CDK or imported.
 */
class InferenceProfileBase extends aws_cdk_lib_1.Resource {
    /**
     * Grants appropriate permissions to use the cross-region inference profile.
     * Does not grant permissions to use the model in the profile.
     */
    grantProfileUsage(grantee) {
        const grant = aws_iam_1.Grant.addToPrincipal({
            grantee: grantee,
            actions: ['bedrock:GetInferenceProfile', 'bedrock:ListInferenceProfiles'],
            resourceArns: [this.inferenceProfileArn],
            scope: this,
        });
        return grant;
    }
}
exports.InferenceProfileBase = InferenceProfileBase;
_a = JSII_RTTI_SYMBOL_1;
InferenceProfileBase[_a] = { fqn: "@cdklabs/generative-ai-cdk-constructs.bedrock.InferenceProfileBase", version: "0.1.309" };
