// @cdklabs/generative-ai-cdk-constructs/lib/cdk-lib/bedrock/inference-profiles/default-prompt-routers.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.PromptRouter = exports.DefaultPromptRouterIdentifier = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
const aws_cdk_lib_1 = require("aws-cdk-lib");
const aws_iam_1 = require("aws-cdk-lib/aws-iam");
const cross_region_inference_profile_1 = require("./../inference-profiles/cross-region-inference-profile");
const models_1 = require("./../models");
/**
 * Represents identifiers for default prompt routers in Bedrock
 */
class DefaultPromptRouterIdentifier {
    constructor(props) {
        (this.promptRouterId = props.promptRouterId), (this.routingModels = props.routingModels);
    }
}
exports.DefaultPromptRouterIdentifier = DefaultPromptRouterIdentifier;
_a = JSII_RTTI_SYMBOL_1;
DefaultPromptRouterIdentifier[_a] = { fqn: "@cdklabs/generative-ai-cdk-constructs.bedrock.DefaultPromptRouterIdentifier", version: "0.1.309" };
/**
 * Anthropic Claude V1 router configuration
 */
DefaultPromptRouterIdentifier.ANTHROPIC_CLAUDE_V1 = new DefaultPromptRouterIdentifier({
    promptRouterId: 'anthropic.claude:1',
    routingModels: [
        models_1.BedrockFoundationModel.ANTHROPIC_CLAUDE_HAIKU_V1_0,
        models_1.BedrockFoundationModel.ANTHROPIC_CLAUDE_3_5_SONNET_V1_0,
    ],
});
/**
 * Meta Llama 3.1 router configuration
 */
DefaultPromptRouterIdentifier.META_LLAMA_3_1 = new DefaultPromptRouterIdentifier({
    promptRouterId: 'meta.llama:1',
    routingModels: [
        models_1.BedrockFoundationModel.META_LLAMA_3_1_8B_INSTRUCT_V1,
        models_1.BedrockFoundationModel.META_LLAMA_3_1_70B_INSTRUCT_V1,
    ],
});
class PromptRouter {
    static fromDefaultId(defaultRouter, region) {
        return new PromptRouter(defaultRouter, region);
    }
    constructor(props, region) {
        this.promptRouterId = props.promptRouterId;
        this.promptRouterArn = aws_cdk_lib_1.Arn.format({
            partition: aws_cdk_lib_1.Aws.PARTITION,
            service: 'bedrock',
            region: region,
            account: aws_cdk_lib_1.Aws.ACCOUNT_ID,
            resource: 'default-prompt-router',
            resourceName: this.promptRouterId,
            arnFormat: aws_cdk_lib_1.ArnFormat.SLASH_RESOURCE_NAME,
        });
        // needed to implement IInvokable
        this.invokableArn = this.promptRouterArn;
        // build inference profiles from routing endpoints
        this.routingEndpoints = props.routingModels.flatMap(model => {
            return cross_region_inference_profile_1.CrossRegionInferenceProfile.fromConfig({
                model: model,
                geoRegion: cross_region_inference_profile_1.REGION_TO_GEO_AREA[region],
            });
        });
    }
    grantInvoke(grantee) {
        // Grant invoke on every model of the router
        this.routingEndpoints.forEach(model => {
            model.grantInvoke(grantee);
        });
        // Grant invoke to the prompt router
        return aws_iam_1.Grant.addToPrincipal({
            grantee,
            actions: ['bedrock:GetPromptRouter', 'bedrock:InvokeModel'],
            resourceArns: [this.promptRouterArn],
        });
    }
}
exports.PromptRouter = PromptRouter;
_b = JSII_RTTI_SYMBOL_1;
PromptRouter[_b] = { fqn: "@cdklabs/generative-ai-cdk-constructs.bedrock.PromptRouter", version: "0.1.309" };
