// @cdklabs/generative-ai-cdk-constructs/lib/cdk-lib/opensearchserverless/analysis-plugins.js
"use strict";
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.TokenFilterType = exports.TokenizerType = exports.CharacterFilterType = void 0;
var CharacterFilterType;
(function (CharacterFilterType) {
    CharacterFilterType["ICU_NORMALIZER"] = "icu_normalizer";
})(CharacterFilterType || (exports.CharacterFilterType = CharacterFilterType = {}));
// Currently we only support Kuromoji and ICU tokenizers.
// Also see the following link for more information regarding supported plugins:
// https://docs.aws.amazon.com/opensearch-service/latest/developerguide/serverless-genref.html#serverless-plugins
var TokenizerType;
(function (TokenizerType) {
    /**
     * Kuromoji tokenizer is used for Japanese text analysis and segmentation
     */
    TokenizerType["KUROMOJI_TOKENIZER"] = "kuromoji_tokenizer";
    /**
     * ICU tokenizer is used for Unicode text segmentation based on UAX #29 rules
     */
    TokenizerType["ICU_TOKENIZER"] = "icu_tokenizer";
    /**
     * Nori tokenizer is used for Korean text analysis and segmentation
     */
    TokenizerType["NORI_TOKENIZER"] = "nori_tokenizer";
})(TokenizerType || (exports.TokenizerType = TokenizerType = {}));
/**
 * TokenFilterType defines the available token filters for text analysis.
 * Token filters process tokens after they have been created by the tokenizer.
 * They can modify, add, or remove tokens based on specific rules.
 */
var TokenFilterType;
(function (TokenFilterType) {
    /**
     * Converts inflected Japanese words to their base form
     */
    TokenFilterType["KUROMOJI_BASEFORM"] = "kuromoji_baseform";
    /**
     * Tags words with their parts of speech in Japanese text analysis
     */
    TokenFilterType["KUROMOJI_PART_OF_SPEECH"] = "kuromoji_part_of_speech";
    /**
     * Reduces Japanese words to their stem form
     */
    TokenFilterType["KUROMOJI_STEMMER"] = "kuromoji_stemmer";
    /**
     * Normalizes CJK width differences by converting all characters to their fullwidth or halfwidth variants
     */
    TokenFilterType["CJK_WIDTH"] = "cjk_width";
    /**
     * Removes Japanese stop words from text
     */
    TokenFilterType["JA_STOP"] = "ja_stop";
    /**
     * Converts all characters to lowercase
     */
    TokenFilterType["LOWERCASE"] = "lowercase";
    /**
     * Applies Unicode folding rules for better text matching
     */
    TokenFilterType["ICU_FOLDING"] = "icu_folding";
    /**
     * Tags words with their parts of speech in Korean text analysis
     */
    TokenFilterType["NORI_PART_OF_SPEECH"] = "nori_part_of_speech";
    /**
     * Converts Korean text to its reading form
     */
    TokenFilterType["NORI_READINGFORM"] = "nori_readingform";
    /**
     * Normalizes Korean numbers to regular Arabic numbers
     */
    TokenFilterType["NORI_NUMBER"] = "nori_number";
})(TokenFilterType || (exports.TokenFilterType = TokenFilterType = {}));
